tck/src/main/java/org/eclipse/microprofile/context/tck/ThreadContextTest.java

```diff
@@ -1426,4 +1413,136 @@ public void currentContextExecutorRunsWithContext() throws InterruptedException,
+
+                return res;
+            });
+
+            CompletableFuture<String> contextCF = labelContext.withContextCapture(noContextCFStage1).thenApplyAsync(res -> {
+                Assert.assertEquals(Label.get(), "default-executor-service-context-test-label-A",
+                        "Context type should be propagated to contextual CompletableFuture.");
+                
+                Assert.assertTrue(Thread.currentThread() instanceof TckThread,
+                        "Current thread should have been created by default executor service");
+                return res;
+            });
+
+            CompletionStage<String> contextCS = labelContext.withContextCapture((CompletionStage<String>)noContextCFStage1).thenApplyAsync(res -> {
+                Assert.assertEquals(Label.get(), "default-executor-service-context-test-label-A",
+                        "Context type should be propagated to contextual CompletionStage.");
+                
+                Assert.assertTrue(Thread.currentThread() instanceof TckThread,
+                        "Current thread should have been created by default executor service");
+                return res;
+            });
+
+            noContextCF.complete("OK");
+            contextCF.get(MAX_WAIT_NS, TimeUnit.NANOSECONDS);
+            contextCS.toCompletableFuture().get(MAX_WAIT_NS, TimeUnit.NANOSECONDS);
+        }
+        finally {
+            // Restore original values
+            Label.set(null);
+            executorService.shutdownNow();
+        }
+    }
+
+    static ContextManager.Builder getContextManagerBuilderIfSupported() {
+        try {
+            return ContextManagerProvider
+                .instance()
+                .getContextManagerBuilder();
+        } catch(UnsupportedOperationException x) {
+            // optional operation: skip
+            throw new SkipException("getContextManagerBuilder not supported", x);
+        }
+    }
+
+    /**
+     * Verify that the lack of a default executor service bans contextualised CompletableFuture/CompletionStage
+     * from running async actions on them.
+     *
+     * @throws ExecutionException indicates test failure
+     * @throws InterruptedException indicates test failure
+     * @throws TimeoutException indicates test failure
+     */
+    @Test
+    public void withoutDefaultExecutorServiceContextCannotInvokeAsyncActions()
+            throws ExecutionException, InterruptedException, TimeoutException {
+        ContextManager contextManager;
+        try {
+            contextManager = getContextManagerBuilderIfSupported()
+                    .addDiscoveredThreadContextProviders()
+                    .withDefaultExecutorService(null)
+                    .build();
+        } catch (SkipException x) {
+            // not supported, use the default and assume that one does not have a default executor service
+            contextManager = ContextManagerProvider.instance().getContextManager();
```

Comment:
I'm not following. I thought the behaviour of `ThreadContext.withContextCapture(CompletionStage)` used to be that `*Async` methods should throw, and should keep on throwing if we can't specify a default executor.

Are you saying that OpenLiberty will make them not throw now?

If that is the case, then I can't check at all whether the `Async` methods throw or not with OpenLiberty because I can't predict whether or not there will be a default executor and so both options would be fine depending on implementation, right?

Reply:
> I'm not following. I thought the behaviour of `ThreadContext.withContextCapture(CompletionStage)` used to be that `*Async` methods should throw

You are correct about what the requirement used to be.

> and should keep on throwing if we can't specify a default executor.

The JavaDoc isn't worded that way.  It states that apart from the `ManagedExecutor.getThreadContext` path, the new completion stages will use "the default executor service if provided by the platform".

> Are you saying that OpenLiberty will make them not throw now?

Yes. OpenLiberty does provide a default executor service and therefore uses it here.  The default executor service happens to not be switchable, but that is beside the point.  While it would be possible to change the spec to prohibit this, I cannot see why we would change it to prohibit because allowing customers to use the *Async methods is the most useful behavior.

> If that is the case, then I can't check at all whether the `Async` methods throw or not with OpenLiberty because I can't predict whether or not there will be a default executor and so both options would be fine depending on implementation, right?

Correct.  Both options are valid for implementations, so a test needs to consider both to be passing results.  It's still worth keeping the test because it will verify there are no failures at other points.